### ISSUE TITLE
Add FirebaseAuth.signInWithEmailAndPassword for Android & iOS

### ISSUE THREAD

**Issue by agatap2** [GitLiveApp/firebase-kotlin-sdk#13]
| Library               | Class                            | Member                                        | Platforms                  |
| :-----------------| :------------------------ | :----------------------------------- | :---------------------- |
| auth | FirebaseAuth | signInWithEmailAndPassword| Android & iOS  |

**Comment by nbransby** [comment]
Hey, I added this on master, firebase-auth version 0.2.0 - can you clone the repo, publish to maven local and see if it works for you? Once you confirm we can publish it to maven central.

**Comment by agatap2** [comment]
Hey, could you tell me how to publish project to local maven?

niedz., 26 kwi 2020 o 22:58 Nicholas Bransby-Williams <
notifications@github.com> napisał(a):

**Comment by nbransby** [comment]
there is a gradle task in the project: `publishToMavenLocal`

then just make sure your client project has `mavenLocal()` in the `repositories` block in its build.gradle

**Comment by agatap2** [comment]
I have added mavenLocal() but your project is still unavailable. I don't
know if it wasn't created after build or if problem is somewhere else. But
I think that you can publish your improvements.

Thanks for quick response.



pon., 27 kwi 2020, 15:12 użytkownik Nicholas Bransby-Williams <
notifications@github.com> napisał:

**Comment by NahroTo** [comment]
This is a duplicate of #12.

**Comment by NahroTo** [comment]
The functionality for this seems to be already available in the master branch. Any reason why this issue hasn't be closed yet?

**Comment by agatap2** [comment]
It's not on maven central and I can't use it.

**Comment by nbransby** [comment]
Waiting for someone to test it works for them locally using the instructions above before I publish it

**Comment by NahroTo** [comment]
I think following standards, the master branch should alawys reflect the latest production version. If it is not tested, I recommend keeping it at a seperate branch.

**Comment by nbransby** [comment]
I have to say I agree! Will make sure this is the case from now on

**Comment by agatap2** [comment]
I was trying to test new functionality, but there is some problem with generating iOS version. 

```
Could not find dev.gitlive:firebase-auth-ios:0.2.0.
Searched in the following locations:
  - file:/C:/Users/Agata/.m2/repository/dev/gitlive/firebase-auth-ios/0.2.0/firebase-auth-ios-0.2.0.module
  - file:/C:/Users/Agata/.m2/repository/dev/gitlive/firebase-auth-ios/0.2.0/firebase-auth-ios-0.2.0.pom
  - https://repo.maven.apache.org/maven2/dev/gitlive/firebase-auth-ios/0.2.0/firebase-auth-ios-0.2.0.module
  - https://repo.maven.apache.org/maven2/dev/gitlive/firebase-auth-ios/0.2.0/firebase-auth-ios-0.2.0.pom
  - https://jcenter.bintray.com/dev/gitlive/firebase-auth-ios/0.2.0/firebase-auth-ios-0.2.0.module
  - https://jcenter.bintray.com/dev/gitlive/firebase-auth-ios/0.2.0/firebase-auth-ios-0.2.0.pom
  - https://dl.google.com/dl/android/maven2/dev/gitlive/firebase-auth-ios/0.2.0/firebase-auth-ios-0.2.0.module
  - https://dl.google.com/dl/android/maven2/dev/gitlive/firebase-auth-ios/0.2.0/firebase-auth-ios-0.2.0.pom
```

When I switch to 0.1.0 version everything works fine.

**Comment by nbransby** [comment]
How does your build gradle look like?

**Comment by agatap2** [comment]
```
plugins {
    id 'com.android.application'
    id 'kotlin-android'
    id 'kotlin-android-extensions'
}

android {
    compileSdkVersion 29
    buildToolsVersion "29.0.3"

    defaultConfig {
        applicationId "com.aramso.multiplatform"
        minSdkVersion 29
        targetSdkVersion 29
        versionCode 1
        versionName "1.0"

        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
    }

    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
    kotlinOptions {
        jvmTarget = '1.8'
    }
    buildFeatures {
        compose true
    }
}

dependencies {

    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
    implementation 'androidx.core:core-ktx:1.2.0'
    implementation 'androidx.appcompat:appcompat:1.1.0'
    implementation 'com.google.android.material:material:1.1.0'
    implementation 'androidx.ui:ui-framework:0.1.0-dev03'
    implementation 'androidx.ui:ui-layout:0.1.0-dev03'
    implementation 'androidx.ui:ui-material:0.1.0-dev03'
    implementation 'androidx.ui:ui-tooling:0.1.0-dev03'
    implementation 'androidx.lifecycle:lifecycle-extensions:2.2.0'

    testImplementation 'junit:junit:4.+'
    androidTestImplementation 'androidx.test.ext:junit:1.1.1'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'

    implementation project(path: ':common')
}
```

**Comment by nbransby** [comment]
cant see your import of dev.gitlive:firebase-auth-ios:0.2.0. here

**Comment by agatap2** [comment]
It's in :common module.

**Comment by nbransby** [comment]
paste that or link if open source project

**Comment by agatap2** [comment]
my new build gradle (error still the same):

```
buildscript {
    ext {
        kotlin_version = "1.3.72"
        coroutine_version = "1.3.2"
        coroutines_play_services = "1.3.5"
        serializer_version = "0.13.0"
        androidx_lifecycle_version = "2.2.0"
        moko_mvvm_version = "0.3.1"
        mockk_version = "1.9.3"
        firebase_auth_version = "0.2.0"
    }
    repositories {
        google()
    }
    dependencies {
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
        classpath "com.android.tools.build:gradle:3.6.3"
        classpath 'com.google.gms:google-services:4.3.3'
    }
}
plugins {
    id 'org.jetbrains.kotlin.multiplatform' version '1.3.72'
}

apply plugin: 'com.android.library'
apply plugin: 'kotlin-android-extensions'
apply plugin: 'org.jetbrains.kotlin.native.cocoapods'

repositories {
    mavenLocal()
    jcenter()
    google()

    maven { url = uri("https://dl.bintray.com/icerockdev/moko") }
}

group 'com.aramso.curtains.backend'
version '0.0.1'

apply plugin: 'maven-publish'

def kotlinPluginId = 'kotlin-multiplatform'
final hasPlugin = project.getPlugins().hasPlugin(kotlinPluginId);
if (hasPlugin) {
    final Plugin plugin = project.getPlugins().getPlugin(kotlinPluginId)
    println 'Plugin already applied - version ' + plugin.properties['kotlinPluginVersion']
} else {
    apply plugin: "kotlin-multiplatform"
}

android {
    compileSdkVersion(29)

    defaultConfig {
        minSdkVersion(21)
        targetSdkVersion(29)
    }

    sourceSets {
        main {
            manifest.srcFile 'src/androidMain/AndroidManifest.xml'
            java.srcDirs = ['src/androidMain/kotlin']
            res.srcDirs = ['src/androidMain/resources']
        }
        test {
            java.srcDirs = ['src/androidTest/kotlin']
            res.srcDirs = ['src/androidTest/resources']
        }
    }

    testOptions.unitTests.includeAndroidResources = true

    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }

    tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
        kotlinOptions {
            jvmTarget = "1.8"
        }
    }
}

kotlin {
    targets {
        final def iOSTarget = System.getenv('SDK_NAME')?.startsWith("iphoneos") \
                              ? presets.iosArm64 : presets.iosX64

        fromPreset(iOSTarget, 'ios') {
            binaries {
                framework('MultiCore')
            }
        }

        fromPreset(presets.android, 'android')
    }
    sourceSets {
        commonMain {
            dependencies {
                implementation kotlin('stdlib-common')

                // COROUTINES
                implementation "org.jetbrains.kotlinx:kotlinx-coroutines-core-common:$coroutine_version"
                implementation "org.jetbrains.kotlinx:kotlinx-coroutines-play-services:$coroutines_play_services"

                // MOKO - MVVM
                implementation "dev.icerock.moko:mvvm:$moko_mvvm_version"

                // SERIALIZATION
                implementation "org.jetbrains.kotlinx:kotlinx-serialization-runtime-common:$serializer_version"

                // FIREBASE AUTH
                implementation "dev.gitlive:firebase-auth:$firebase_auth_version"
            }
        }
        commonTest {
            dependencies {
                implementation kotlin('test-common')
                implementation kotlin('test-annotations-common')

                implementation "io.mockk:mockk:$mockk_version"
            }
        }
        androidMain {
            dependencies {
                implementation kotlin('stdlib')

                implementation "androidx.lifecycle:lifecycle-extensions:$androidx_lifecycle_version"
                implementation "org.jetbrains.kotlin:kotlin-stdlib-common:$kotlin_version"

                // COROUTINES
                implementation "org.jetbrains.kotlinx:kotlinx-coroutines-android:$coroutine_version"

                // SERIALIZATION
                implementation "org.jetbrains.kotlinx:kotlinx-serialization-runtime:$serializer_version"
            }
        }
        androidTest {
            dependencies {
                implementation kotlin('test')
                implementation kotlin('test-junit')
            }
        }
        iosMain {
        }
        iosTest {
        }
    }
}

configurations {
    compileClasspath
}

task packForXcode(type: Sync) {
    final File frameworkDir = new File(buildDir, "xcode-frameworks")
    final String mode = project.findProperty("XCODE_CONFIGURATION")?.toUpperCase() ?: 'DEBUG'
    final def framework = kotlin.targets.ios.binaries.getFramework("MultiCore", mode)

    inputs.property "mode", mode
    dependsOn framework.linkTask

    from { framework.outputFile.parentFile }
    into frameworkDir

    doLast {
        new File(frameworkDir, 'gradlew').with {
            text = "#!/bin/bash\nexport 'JAVA_HOME=${System.getProperty("java.home")}'\ncd '${rootProject.rootDir}'\n./gradlew \$@\n"
            setExecutable(true)
        }
    }
}

tasks.build.dependsOn packForXcode
```

**Comment by agatap2** [comment]
New build gradle - error the same:

```
buildscript {
    ext {
        kotlin_version = "1.3.72"
        coroutine_version = "1.3.2"
        coroutines_play_services = "1.3.5"
        serializer_version = "0.13.0"
        androidx_lifecycle_version = "2.2.0"
        moko_mvvm_version = "0.3.1"
        mockk_version = "1.9.3"
        firebase_auth_version = "0.2.0"
    }
    repositories {
        google()
    }
    dependencies {
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
        classpath "com.android.tools.build:gradle:3.6.3"
        classpath 'com.google.gms:google-services:4.3.3'
    }
}
plugins {
    id 'org.jetbrains.kotlin.multiplatform' version '1.3.72'
}

apply plugin: 'com.android.library'
apply plugin: 'kotlin-android-extensions'
apply plugin: 'org.jetbrains.kotlin.native.cocoapods'

repositories {
    mavenLocal()
    jcenter()
    google()

    maven { url = uri("https://dl.bintray.com/icerockdev/moko") }
}

group 'com.aramso.curtains.backend'
version '0.0.1'

apply plugin: 'maven-publish'

def kotlinPluginId = 'kotlin-multiplatform'
final hasPlugin = project.getPlugins().hasPlugin(kotlinPluginId);
if (hasPlugin) {
    final Plugin plugin = project.getPlugins().getPlugin(kotlinPluginId)
    println 'Plugin already applied - version ' + plugin.properties['kotlinPluginVersion']
} else {
    apply plugin: "kotlin-multiplatform"
}

android {
    compileSdkVersion(29)

    defaultConfig {
        minSdkVersion(21)
        targetSdkVersion(29)
    }

    sourceSets {
        main {
            manifest.srcFile 'src/androidMain/AndroidManifest.xml'
            java.srcDirs = ['src/androidMain/kotlin']
            res.srcDirs = ['src/androidMain/resources']
        }
        test {
            java.srcDirs = ['src/androidTest/kotlin']
            res.srcDirs = ['src/androidTest/resources']
        }
    }

    testOptions.unitTests.includeAndroidResources = true

    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }

    tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
        kotlinOptions {
            jvmTarget = "1.8"
        }
    }
}

kotlin {
    targets {
        final def iOSTarget = System.getenv('SDK_NAME')?.startsWith("iphoneos") \
                              ? presets.iosArm64 : presets.iosX64

        fromPreset(iOSTarget, 'ios') {
            binaries {
                framework('MultiCore')
            }
        }

        fromPreset(presets.android, 'android')
    }
    sourceSets {
        commonMain {
            dependencies {
                implementation kotlin('stdlib-common')

                // COROUTINES
                implementation "org.jetbrains.kotlinx:kotlinx-coroutines-core-common:$coroutine_version"
                implementation "org.jetbrains.kotlinx:kotlinx-coroutines-play-services:$coroutines_play_services"

                // MOKO - MVVM
                implementation "dev.icerock.moko:mvvm:$moko_mvvm_version"

                // SERIALIZATION
                implementation "org.jetbrains.kotlinx:kotlinx-serialization-runtime-common:$serializer_version"

                // FIREBASE AUTH
                implementation "dev.gitlive:firebase-auth:$firebase_auth_version"
            }
        }
        commonTest {
            dependencies {
                implementation kotlin('test-common')
                implementation kotlin('test-annotations-common')

                implementation "io.mockk:mockk:$mockk_version"
            }
        }
        androidMain {
            dependencies {
                implementation kotlin('stdlib')

                implementation "androidx.lifecycle:lifecycle-extensions:$androidx_lifecycle_version"
                implementation "org.jetbrains.kotlin:kotlin-stdlib-common:$kotlin_version"

                // COROUTINES
                implementation "org.jetbrains.kotlinx:kotlinx-coroutines-android:$coroutine_version"

                // SERIALIZATION
                implementation "org.jetbrains.kotlinx:kotlinx-serialization-runtime:$serializer_version"
            }
        }
        androidTest {
            dependencies {
                implementation kotlin('test')
                implementation kotlin('test-junit')
            }
        }
        iosMain {
        }
        iosTest {
        }
    }
}

configurations {
    compileClasspath
}

task packForXcode(type: Sync) {
    final File frameworkDir = new File(buildDir, "xcode-frameworks")
    final String mode = project.findProperty("XCODE_CONFIGURATION")?.toUpperCase() ?: 'DEBUG'
    final def framework = kotlin.targets.ios.binaries.getFramework("MultiCore", mode)

    inputs.property "mode", mode
    dependsOn framework.linkTask

    from { framework.outputFile.parentFile }
    into frameworkDir

    doLast {
        new File(frameworkDir, 'gradlew').with {
            text = "#!/bin/bash\nexport 'JAVA_HOME=${System.getProperty("java.home")}'\ncd '${rootProject.rootDir}'\n./gradlew \$@\n"
            setExecutable(true)
        }
    }
}

tasks.build.dependsOn packForXcode
```

**Comment by nbransby** [comment]
Have a look in `C:/Users/Agata/.m2/repository/dev/gitlive/` should be in there if you have publishedtoMavenLocal successfull .

But that a window path, the ios module won't be built on windows but also you cant build a ios project on windows anyway

**Comment by agatap2** [comment]
It's not there but now I know why. I thought it would be built on Windows, even if I can't use it there.

**Comment by agatap2** [comment]
I have written extension function in common module:
```
suspend fun FirebaseAuth.authenticate(
    email: String,
    password: String
): AuthResult? {
    return signInWithEmailAndPassword(email, password)
}
```
and I'm trying to call it in viewmodelscope like this:
```Firebase.auth.authenticate(email, password)```

but I'm getting such error:
`java.lang.IllegalStateException: Default FirebaseApp is not initialized in this process com.... Make sure to call FirebaseApp.initializeApp(Context) first.`

Where should I call it?

**Comment by agatap2** [comment]
Tested, function is working correctly.

**Comment by vladimirpetrovski** [comment]
I can also confirm that this feature works well on Android using locally published artefacts for version 0.2.0.
